File: Python/Two Sum.py
from typing import List


class Solution:
    """
    Time:   O(n)
    Memory: O(n)
    """

    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        index = {}
        for i, num in enumerate(numbers):
            if target - num in index:
                return [index[target - num], i]
            index[num] = i

File: Java/Two Sum.java
import java.util.HashMap;

class Solution {

    /**
     * Time:   O(n)
     * Memory: O(n)
     */
    public int[] twoSum(int[] numbers, int target) {
        HashMap<Integer, Integer> hash = new HashMap<>();

        for (int i = 0; i < numbers.length; i++) {
            Integer diff = target - numbers[i];
            if (hash.containsKey(diff))
                return new int[]{hash.get(diff), i};
            hash.put(numbers[i], i);
        }

        return null;
    }
}
